### PR TITLE
Fix up "linux*" terminfo entries, too.

### PR DESCRIPTION
**Pull request by jdebp**
The terminfo entry for `linux` only advertises 8 colours, and its `setaf` and `setab` capabilities only work correctly with colours 0 to 7, but nvim tries to make it display 16 colours anyway, resulting in erroneous SGR control sequences for colours 8 and above.

The Linux kernel terminal emulator itself has actually understood the 256-colour control sequences [since version 4.8](https://github.com/torvalds/linux/commit/cec5b2a97a11ade56a701e83044d0a2a984c67b4) (at least ignoring them [well before that](https://github.com/torvalds/linux/commit/3415097ff0529eac264b8ccfa06572871e45c090)) and the 16-colour control sequences since version 4.9 ([foreground](https://github.com/torvalds/linux/commit/cc67dc28b33917227214a692534d817d727fb934) and [background](https://github.com/torvalds/linux/commit/fadb4244085cd04fd9c8b3a4b3bc161f506431f3)).

Thus we apply the same terminfo fixup to `linux*` as we apply for` *xterm*` and` *256*`, to emit the 16-colour and 256-colour control sequences even if terminfo's `setaf` and `setab` do not advertise them.

### CODE REVIEW

**Comment by justinmk** [comment]
Hmm, ok. This won't help #5538--rather, it will make that a "wontfix"? cc @f3k7c

Linux 4.8 is pretty recent. What can users on older kernels do to get a usable display?

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -1016,8 +1016,12 @@ static void fix_terminfo(TUIData *data)
   "\x1b[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m"
 
   if ((colorterm && strstr(colorterm, "256"))
+      || STARTS_WITH(term, "linux")
       || strstr(term, "256")
       || strstr(term, "xterm")) {
+    // Linux has supported the 256-colour control sequences since 4.8, but its
+    // terminfo entry as of 2016-11-27 only advertises 8 colour support.
+    // This also fixes up the linux-16color terminfo entry as a bonus.
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: The details can live in the commit message. Let's not give more than 1 line to this comment.

    // Linux 4.8+ supports 256-colors, but terminfo only claims 8.

**Comment by jdebp** [comment]
Well it definitely fixes the colours when run with the `linux` terminal type, and the black-on-black parts do go away.

**Comment by justinmk** [comment]
If it improves things then sure let's merge it. It won't make #5538 worse, I think.

**Review by justinmk — CHANGES_REQUESTED** [review]
see comment

**Comment by jdebp** [comment]
The cause of the black on black, incidentally, is that nvim ends up trying to change colour by emitting SGR control sequences of the form `CSI` `[` `4` `1` `1` `m` (to change to colour number 11), which of course do not work, leaving the prior background colour still in effect.  This was a fairly tedious exercise in the use of `strace` to discover this.  (-:

There are a few problems that all have this as their root, including `hlsearch` not actually highlighting anything on a `linux` type terminal.  This now happens.

File: src/nvim/tui/tui.c (hunk 1 above)
Comment (jdebp) [inline src/nvim/tui/tui.c]: Feel free to do that if you want.

**Comment by jdebp** [comment]
The parameter is not in the range `40` to `47`, which is what the vanilla terminfo capability is designed to (only) produce.  `411` has no defined meaning.  That's not the way to specify colour number 11.  (-:

**Comment by justinmk** [comment]
Thanks. I was staring at the `CSI Pm m` section of [ctlseqs](http://invisible-island.net/xterm/ctlseqs/ctlseqs.html), and I see `Ps = 4 ?` allows 0-9 for the "?". I used the wrong words, as usual :)

**Comment by justinmk** [comment]
merged

**Comment by jdebp** [comment]
Note that I did change the comment.

**Comment by jamessan** [comment]
> trying to change colour by emitting SGR control sequences of the form CSI 4 1 1 m (to change to colour number 11)

Was that supposed to be <kbd>CSI</kbd><kbd>41</kbd><kbd>;</kbd><kbd>1</kbd><kbd>m</kbd> (bold red)?

**Comment by jdebp** [comment]
No.  It's the straightforward consequence of a very simple `\E[4%p1%dm` terminfo capability, which goes wrong when the colour index exceeds 7.  The more complex terminfo capability that nvim substitutes does the right thing for colour indices 7 and up, emitting the 16-colour and 256-colour SGR sequences.

**Pull request closed without merging**